File: backend/routers/locations.py
"""Location Management routes (Regions, Cities, Districts)"""
from fastapi import APIRouter, HTTPException, Query, Depends, Header
from typing import List, Optional

from database import db
from models import Region, City, District, LocationCreate
from auth import require_auth, resolve_tenant_filter, resolve_tenant_id

router = APIRouter(tags=["Locations"])


# ============== REGIONS ==============

@router.post("/regions", response_model=Region)
async def create_region(input: LocationCreate, user: dict = Depends(require_auth), x_tenant_id: Optional[str] = Header(None)):
    tenant_id = resolve_tenant_id(user, x_tenant_id)
    region = Region(name=input.name, tenant_id=tenant_id)
    doc = region.model_dump()
    doc["created_at"] = doc["created_at"].isoformat()
    await db.regions.insert_one(doc)
    return region


@router.get("/regions", response_model=List[Region])
async def get_regions(user: dict = Depends(require_auth)):
    regions = await db.regions.find({}, {"_id": 0}).to_list(100)
    return regions


@router.delete("/regions/{region_id}")
async def delete_region(region_id: str, user: dict = Depends(require_auth), x_tenant_id: Optional[str] = Header(None)):
    query = {"id": region_id, **resolve_tenant_filter(user, x_tenant_id)}
    result = await db.regions.delete_one(query)
    if result.deleted_count == 0:
        raise HTTPException(status_code=404, detail="Region not found")
    return {"status": "deleted"}


# ============== CITIES ==============

@router.post("/cities", response_model=City)
async def create_city(input: LocationCreate, user: dict = Depends(require_auth), x_tenant_id: Optional[str] = Header(None)):
    if not input.parent_id:
        raise HTTPException(status_code=400, detail="parent_id (region_id) is required")
    tenant_id = resolve_tenant_id(user, x_tenant_id)
    city = City(name=input.name, region_id=input.parent_id, tenant_id=tenant_id)
    doc = city.model_dump()
    doc["created_at"] = doc["created_at"].isoformat()
    await db.cities.insert_one(doc)
    return city


@router.get("/cities", response_model=List[City])
async def get_cities(region_id: Optional[str] = Query(None), user: dict = Depends(require_auth)):
    query = {}
    if region_id:
        query["region_id"] = region_id
    cities = await db.cities.find(query, {"_id": 0}).to_list(100)
    return cities


@router.delete("/cities/{city_id}")
async def delete_city(city_id: str, user: dict = Depends(require_auth), x_tenant_id: Optional[str] = Header(None)):
    query = {"id": city_id, **resolve_tenant_filter(user, x_tenant_id)}
    result = await db.cities.delete_one(query)
    if result.deleted_count == 0:
        raise HTTPException(status_code=404, detail="City not found")
    return {"status": "deleted"}


# ============== DISTRICTS ==============

@router.post("/districts", response_model=District)
async def create_district(input: LocationCreate, user: dict = Depends(require_auth), x_tenant_id: Optional[str] = Header(None)):
    if not input.parent_id:
        raise HTTPException(status_code=400, detail="parent_id (city_id) is required")
    tenant_id = resolve_tenant_id(user, x_tenant_id)
    district = District(name=input.name, city_id=input.parent_id, tenant_id=tenant_id)
    doc = district.model_dump()
    doc["created_at"] = doc["created_at"].isoformat()
    await db.districts.insert_one(doc)
    return district


@router.get("/districts", response_model=List[District])
async def get_districts(city_id: Optional[str] = Query(None), user: dict = Depends(require_auth)):
    query = {}
    if city_id:
        query["city_id"] = city_id
    districts = await db.districts.find(query, {"_id": 0}).to_list(100)
    return districts


@router.delete("/districts/{district_id}")
async def delete_district(district_id: str, user: dict = Depends(require_auth), x_tenant_id: Optional[str] = Header(None)):
    query = {"id": district_id, **resolve_tenant_filter(user, x_tenant_id)}
    result = await db.districts.delete_one(query)
    if result.deleted_count == 0:
        raise HTTPException(status_code=404, detail="District not found")
    return {"status": "deleted"}


# ============== HIERARCHY ==============

@router.get("/hierarchy")
async def get_hierarchy(user: dict = Depends(require_auth), x_tenant_id: Optional[str] = Header(None)):
    """Get complete location hierarchy — regions/cities/districts are global, stores are tenant-scoped"""
    regions = await db.regions.find({}, {"_id": 0}).to_list(100)
    cities = await db.cities.find({}, {"_id": 0}).to_list(500)
    districts = await db.districts.find({}, {"_id": 0}).to_list(1000)
    stores = await db.stores.find(resolve_tenant_filter(user, x_tenant_id), {"_id": 0}).to_list(500)

    hierarchy = []
    for region in regions:
        region_data = {"id": region["id"], "name": region["name"], "type": "region", "cities": []}
        region_cities = [c for c in cities if c.get("region_id") == region["id"]]
        for city in region_cities:
            city_data = {"id": city["id"], "name": city["name"], "type": "city", "districts": []}
            city_districts = [d for d in districts if d.get("city_id") == city["id"]]
            for district in city_districts:
                district_stores = [s for s in stores if s.get("district_id") == district["id"]]
                city_data["districts"].append({
                    "id": district["id"],
                    "name": district["name"],
                    "type": "district",
                    "store_count": len(district_stores),
                })
            region_data["cities"].append(city_data)
        hierarchy.append(region_data)

    return {"hierarchy": hierarchy}

File: backend-node/routes/locations.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const { getDB } = require('../db');
const { requireAuth } = require('../middleware/auth');

const router = express.Router();

// REGIONS
router.post('/regions', requireAuth, async (req, res) => {
  const db = getDB();
  const doc = { id: uuidv4(), name: req.body.name, created_at: new Date().toISOString() };
  await db.collection('regions').insertOne(doc);
  const { _id, ...out } = doc;
  res.json(out);
});

router.get('/regions', requireAuth, async (req, res) => {
  const db = getDB();
  const regions = await db.collection('regions').find({}, { projection: { _id: 0 } }).toArray();
  res.json(regions);
});

router.delete('/regions/:id', requireAuth, async (req, res) => {
  const db = getDB();
  const result = await db.collection('regions').deleteOne({ id: req.params.id });
  if (result.deletedCount === 0) return res.status(404).json({ detail: 'Region bulunamadı' });
  res.json({ status: 'deleted' });
});

// CITIES
router.post('/cities', requireAuth, async (req, res) => {
  const { name, parent_id } = req.body;
  if (!parent_id) return res.status(400).json({ detail: 'parent_id (region_id) gerekli' });
  const db = getDB();
  const doc = { id: uuidv4(), name, region_id: parent_id, created_at: new Date().toISOString() };
  await db.collection('cities').insertOne(doc);
  const { _id, ...out } = doc;
  res.json(out);
});

router.get('/cities', requireAuth, async (req, res) => {
  const db = getDB();
  const query = req.query.region_id ? { region_id: req.query.region_id } : {};
  const cities = await db.collection('cities').find(query, { projection: { _id: 0 } }).toArray();
  res.json(cities);
});

router.delete('/cities/:id', requireAuth, async (req, res) => {
  const db = getDB();
  const result = await db.collection('cities').deleteOne({ id: req.params.id });
  if (result.deletedCount === 0) return res.status(404).json({ detail: 'City bulunamadı' });
  res.json({ status: 'deleted' });
});

// DISTRICTS
router.post('/districts', requireAuth, async (req, res) => {
  const { name, parent_id } = req.body;
  if (!parent_id) return res.status(400).json({ detail: 'parent_id (city_id) gerekli' });
  const db = getDB();
  const doc = { id: uuidv4(), name, city_id: parent_id, created_at: new Date().toISOString() };
  await db.collection('districts').insertOne(doc);
  const { _id, ...out } = doc;
  res.json(out);
});

router.get('/districts', requireAuth, async (req, res) => {
  const db = getDB();
  const query = req.query.city_id ? { city_id: req.query.city_id } : {};
  const districts = await db.collection('districts').find(query, { projection: { _id: 0 } }).toArray();
  res.json(districts);
});

router.delete('/districts/:id', requireAuth, async (req, res) => {
  const db = getDB();
  const result = await db.collection('districts').deleteOne({ id: req.params.id });
  if (result.deletedCount === 0) return res.status(404).json({ detail: 'District bulunamadı' });
  res.json({ status: 'deleted' });
});

// HIERARCHY
router.get('/hierarchy', requireAuth, async (req, res) => {
  const db = getDB();
  const [regions, cities, districts, stores] = await Promise.all([
    db.collection('regions').find({}, { projection: { _id: 0 } }).toArray(),
    db.collection('cities').find({}, { projection: { _id: 0 } }).toArray(),
    db.collection('districts').find({}, { projection: { _id: 0 } }).toArray(),
    db.collection('stores').find({}, { projection: { _id: 0 } }).toArray(),
  ]);

  const hierarchy = regions.map(region => ({
    id: region.id, name: region.name, type: 'region',
    cities: cities.filter(c => c.region_id === region.id).map(city => ({
      id: city.id, name: city.name, type: 'city',
      districts: districts.filter(d => d.city_id === city.id).map(district => ({
        id: district.id, name: district.name, type: 'district',
        store_count: stores.filter(s => s.district_id === district.id).length
      }))
    }))
  }));

  res.json({ hierarchy });
});

module.exports = router;
